sdk/python/feast/feature_store.py

```diff
@@ -153,26 +160,131 @@ def get_historical_features(
+            >>> print(online_response_dict)
+            {'sales:daily_transactions': [1.1,1.2], 'sales:customer_id': [0,1]}
+        """
 
-def _get_requested_feature_views(
+        response = self._get_online_features(
+            feature_refs=feature_refs,
+            entity_rows=_infer_online_entity_rows(entity_rows),
+            project=self.config.project,
+        )
+
+        return OnlineResponse(response)
+
+    def _get_online_features(
+        self,
+        entity_rows: List[GetOnlineFeaturesRequestV2.EntityRow],
+        feature_refs: List[str],
+        project: str,
+    ) -> GetOnlineFeaturesResponse:
+
+        provider = self._get_provider()
+
+        entity_keys = []
+        result_rows: List[GetOnlineFeaturesResponse.FieldValues] = []
+
+        for row in entity_rows:
+            entity_keys.append(_entity_row_to_key(row))
+            result_rows.append(_entity_row_to_field_values(row))
+
+        registry = self._get_registry()
+        all_feature_views = registry.list_feature_views(project=self.config.project)
+
+        grouped_refs = _group_refs(feature_refs, all_feature_views)
+        for table, requested_features in grouped_refs:
+            read_rows = provider.online_read(
+                project=project, table=table, entity_keys=entity_keys,
+            )
+            for row_idx, read_row in enumerate(read_rows):
+                row_ts, feature_data = read_row
+                result_row = result_rows[row_idx]
+
+                if feature_data is None:
+                    for feature_name in requested_features:
+                        feature_ref = f"{table.name}:{feature_name}"
+                        result_row.statuses[
+                            feature_ref
+                        ] = GetOnlineFeaturesResponse.FieldStatus.NOT_FOUND
+                else:
+                    for feature_name in feature_data:
+                        feature_ref = f"{table.name}:{feature_name}"
+                        if feature_name in requested_features:
+                            result_row.fields[feature_ref].CopyFrom(
+                                feature_data[feature_name]
+                            )
+                            result_row.statuses[
+                                feature_ref
+                            ] = GetOnlineFeaturesResponse.FieldStatus.PRESENT
+
+        return GetOnlineFeaturesResponse(field_values=result_rows)
+
+
+def _entity_row_to_key(row: GetOnlineFeaturesRequestV2.EntityRow) -> EntityKeyProto:
+    names, values = zip(*row.fields.items())
+    return EntityKeyProto(entity_names=names, entity_values=values)
+
+
+def _entity_row_to_field_values(
+    row: GetOnlineFeaturesRequestV2.EntityRow,
+) -> GetOnlineFeaturesResponse.FieldValues:
+    result = GetOnlineFeaturesResponse.FieldValues()
+    for k in row.fields:
+        result.fields[k].CopyFrom(row.fields[k])
+        result.statuses[k] = GetOnlineFeaturesResponse.FieldStatus.PRESENT
```

Comment:
Is there a reason for the hardcoding here?

Reply:
If I read the previous implementation correctly, we return feature values as well as entity field values in `FieldValues`. This code populates entity field values, and unlike features, these are always "present" since they're part of the request.